File: migrations/initialize.ts
import { BN } from "@coral-xyz/anchor";
import { PublicKey } from "@solana/web3.js";
import { getKeypairFromFile } from "@solana-developers/helpers";

import { buildAndSendTx, getConfig, prettify } from "../common/utils";
import { IxBuilder } from "../common/ixBuilder";

async function main() {
  const args = process.argv.slice(2);
  const [
    deployerPath,
    adminAddress,
    tournamentNonceStr,
    platformFeeStr,
    minTeamsStr,
    maxTeamsStr,
    falsePrecisionStr,
    maxOrganizerFeeStr,
    consensusRateStr,
    platformWalletAddress,
    nomeMintAddress
  ] = args;

  const deployer = await getKeypairFromFile(deployerPath);
  const admin = new PublicKey(adminAddress);
  const nomeMint = new PublicKey(nomeMintAddress);
  const platformWallet = new PublicKey(platformWalletAddress);

  console.log(`Deployer: ${deployer.publicKey.toBase58()}`);

  const ixBuilder = new IxBuilder();
  const initializeIx = await ixBuilder.initializeIx(deployer.publicKey, {
    tournamentNonce: parseInt(tournamentNonceStr),
    platformFee: new BN(platformFeeStr),
    minTeams: parseInt(minTeamsStr),
    maxTeams: parseInt(maxTeamsStr),
    falsePrecision: parseFloat(falsePrecisionStr),
    consensusRate: parseFloat(consensusRateStr),
    maxOrganizerFee: new BN(maxOrganizerFeeStr),
    admin,
    platformWallet,
    nomeMint,
    verifierAddresses: []
  });

  const txSignature = await buildAndSendTx([initializeIx], [deployer]);
  console.log("Initialize Genome tx:", txSignature);

  const config = await getConfig();
  console.log(`GenomeConfig: ${prettify(config)}`);
}

main()
  .then(() => process.exit(0))
  .catch((error) => {
    console.error(error);
    process.exit(1);
  });

File: migrations/initialize.rs
import * as fs from "fs";
import * as anchor from "@coral-xyz/anchor";
import { PublicKey, Transaction } from "@solana/web3.js";
import { utf8 } from "@coral-xyz/anchor/dist/cjs/utils/bytes";
import { getKeypairFromFile } from "@solana-developers/node-helpers";
import { GenomeContract } from "../target/types/genome_contract";

const GENOME_ROOT = utf8.encode("genome");
const CONFIG = utf8.encode("config");

async function main() {
  const deployerPath = process.argv[2];
  const keysPath = process.argv[3];

  const deployer = await getKeypairFromFile(deployerPath);
  const admin = await getKeypairFromFile(keysPath + "/admin.json");

  console.log("deployer: ", deployer.publicKey.toBase58());
  console.log("admin: ", admin.publicKey.toBase58());

  if (!fs.existsSync(keysPath)) {
    fs.mkdirSync(keysPath);
  }

  const provider = anchor.AnchorProvider.env();
  anchor.setProvider(provider);
  const program = anchor.workspace.GenomeContract as anchor.Program<GenomeContract>;

  const [configPda, configBump] = PublicKey.findProgramAddressSync(
    [GENOME_ROOT, CONFIG],
    program.programId
  );
  console.log("Config PDA:", configPda.toBase58());

  const configData = {
    admin: admin.publicKey,
    tournamentNonce: 0,
    platformFee: new anchor.BN(10),
    platformWallet: admin.publicKey,
    minEntryFee: new anchor.BN(10),
    minSponsorPool: new anchor.BN(500),
    minTeams: 2,
    maxTeams: 20,
    maxOrganizerRoyalty: new anchor.BN(5000),
    assetMint: admin.publicKey,
    trustedAgents: [admin.publicKey],
    consensusThreshold: 1,
  };

  const tx: string = await program.methods
    .initialize(configData)
    .accounts({
      admin: deployer.publicKey,
      config: configPda,
      systemProgram: anchor.web3.SystemProgram.programId,
    })
    .signers([deployer])
    .rpc();

  console.log("Initialize tx:", tx);
}

main()
  .then(() => process.exit(0))
  .catch((error) => {
    console.error(error);
    process.exit(1);
  });
